optional_plugins/varianter_serialized/avocado_varianter_serialized/__init__.py

```diff
@@ -0,0 +1,135 @@
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright: Red Hat Inc. 2018
+# Authors: Amador Pahim <apahim@redhat.com>
+
+import json
+import sys
+
+from six import iteritems
+
+from avocado.core import exit_codes
+from avocado.core import tree
+from avocado.core import varianter
+from avocado.core.output import LOG_UI
+from avocado.core.plugin_interfaces import CLI
+from avocado.core.plugin_interfaces import Varianter
+
+
+class VarianterSerializedCLI(CLI):
+
+    """
+    Serialized based Varianter options
+    """
+
+    name = 'serialized'
+    description = "Serialized based Varianter options for the 'run' subcommand"
+
+    def configure(self, parser):
+
+        for name in ("run", "variants"):  # intentionally omitting "multiplex"
+            subparser = parser.subcommands.choices.get(name, None)
+            if subparser is None:
+                continue
+            sparser = subparser.add_argument_group('json serialized varianter '
+                                                   'options')
+            sparser.add_argument('--load-variants', default=None,
+                                 help=('Load the Variants from a JSON '
+                                       'serialized file'))
+
+    def run(self, args):
+        pass
+
+
+class VarianterSerialized(Varianter):
+
+    """
+    Processes the serialized file into variants
+    """
+
+    name = 'serialized'
+    description = "JSON Serialized Varianter"
+
+    def initialize(self, args):
+        self.variants = None
+        load_variants = getattr(args, "load_variants", None)
+
+        if load_variants is None:
+            return
+        try:
+            with open(load_variants, 'r') as var_file:
+                self.variants = varianter.Varianter(state=json.load(var_file))
+        except IOError:
+            LOG_UI.error("JSON serialized file '%s' could not be found or "
+                         "is not readable", load_variants)
+            if args.subcommand == 'run':
+                sys.exit(exit_codes.AVOCADO_JOB_FAIL)
+            else:
+                sys.exit(exit_codes.AVOCADO_FAIL)
+
+    def __iter__(self):
+        if self.variants is None:
+            return ''
+
+        return self.variants.itertests()
+
+    def __len__(self):
+        if self.variants is None:
+            return 0
```

Comment:
The same here. I'd suggest `RuntimeError` but if you think it's a valid use case returning `0` is probably appropriate.

Reply:
We have to return a valid length here. `0` is actually appropriate.

Reply:
It's hard to decide what is appropriate, because the `len` is unknown before calling `initialize`. Even 0 variants is a valid number of variants so calling `len(...)` before `initialize` should IMO raise the exception, because otherwise the number of variants might change. Do you have some use-case for `len` before `initialize`?